http/logical.go

```diff
@@ -200,6 +200,61 @@ func handleLogicalInternal(core *vault.Core, injectDataIntoTopLevel bool) http.H
 			// requests for these paths always go to the token NS
 			case "auth/token/lookup-self", "auth/token/renew-self", "auth/token/revoke-self":
 				r = r.WithContext(newCtx)
+
+			// For the following operations, we can set the proper namespace context
+			// using the token's embedded nsID if a relative path was provided. Since
+			// this is done at the HTTP layer, the operation will still be gated by
+			// ACLs.
+			case "auth/token/renew", "auth/token/revoke", "auth/token/revoke-orphan":
+				token, ok := req.Data["token"]
+				// If the token is not present (e.g. a bad request), break out and let the backend
+				// handle the error
+				if !ok {
+					break
+				}
+				_, nsID := namespace.SplitIDFromString(token.(string))
+				if nsID != "" {
+					ns, err := vault.NamespaceByID(newCtx, nsID, core)
+					if err != nil {
+						core.Logger().Warn("error looking up namespace from the token's namespace ID", "error", err)
+						respondError(w, http.StatusInternalServerError, err)
+						return
+					}
+					if ns != nil {
+						newCtx = namespace.ContextWithNamespace(newCtx, ns)
+					}
+				}
+				r = r.WithContext(newCtx)
```

Comment:
If the ns != nil check above fails, you'll end up setting the request ns to the authorizing token ns instead of the original request ns, which I think is actually not what you want.

Reply:
Fixed